### ISSUE TITLE
Antialiasing Issues / Stairing

### ISSUE THREAD

**Issue by CodeGod911** [MarvinKlein1508/SignaturePad#39]
Because of the new way the lines are drawn stairing occurs due to saturation of antialiased areas over the drawtime. This is due to repeated drawing of already drawn lines and the antialiased areas around the lines stacking every render.

Ill look into it when i find time.

After minimal drawtime <1sec
![image](https://github.com/MarvinKlein1508/SignaturePad/assets/58259484/0306621a-1441-464c-a1ea-725fa6491b34)

After drawtime 2sec<
 
![image](https://github.com/MarvinKlein1508/SignaturePad/assets/58259484/16af607a-037f-47af-92d9-1b61882747b2)

**Comment by JosipNizetic** [comment]
Hello,


Just regarding Draw Function Point. 


```
   // draw a bunch of quadratics, using the average of two points as the control point

    var i;
    for (i = 1; i < points.length - 2; i++) {
        var c = (points[i].x + points[i + 1].x) / 2,
            d = (points[i].y + points[i + 1].y) / 2;
        ctx.quadraticCurveTo(points[i].x, points[i].y, c, d);
    }

```
maybe just wrong order in quadraticCurveTo

```

    var i;
    for (i = 0; i < points.length - 2; i++) {
        var c = (points[i].x + points[i + 1].x) / 2,
            d = (points[i].y + points[i + 1].y) / 2;
        ctx.quadraticCurveTo( c, d, points[i + 1].x, points[i + 1].y);
    }
```

I am more c+ and desktop C# so have problem with testing, but this is only thing it pop up to me.

**Comment by CodeGod911** [comment]
i once texted that and it produced what i call the "barbed wire effect" :D
![image](https://github.com/MarvinKlein1508/SignaturePad/assets/58259484/133b2baf-588e-4bb6-9a67-14f218f41bf4)

**Comment by CodeGod911** [comment]
@MarvinKlein1508 i have fix that works but one culprit that stands in the way im just not able to avoid. its blazor redrawing the canvas as image every time sigpad.finish is executed any idea?

**Comment by JosipNizetic** [comment]
Hello,

Yes at the end your way was better:

```
 for (i = 1; i < points.length - 2; i++) {
        var c = (points[i].x + points[i + 1].x) / 2,
            d = (points[i].y + points[i + 1].y) / 2;
        ctx.quadraticCurveTo(points[i].x, points[i].y, c, d);
```

Regarding renderer problem i changed renderer fuction:

```
    render(finished) {
        if (this._element.width != this._element.clientWidth) {
            this._element.width = this._element.clientWidth;
        }
       
            var ctx = this._element.getContext("2d");
            this._applyOptions(ctx);
			ctx.clearRect(0, 0,  this._element.clientWidth ,this._element.clientHeight );
            this._mousePosis.filter((stroke) => stroke.length > 0).forEach((stroke) => drawPoints(ctx, stroke, finished));
    }
```

and now this is result

https://github.com/MarvinKlein1508/SignaturePad/assets/63609051/765279c2-1444-48e5-9778-8c35990e5f79

**Comment by CodeGod911** [comment]
Unfortunately this was the way it was before @MarvinKlein1508 found out that loading signatures and setting backgroundimages stopped working because of that line...

**Comment by JosipNizetic** [comment]
OK. 

Yes, I never use backgrounds, so that scenario slipped by me.
Maybe some combination with ctx restore and save combination. To save before mouse down and then writing on it.
https://stackoverflow.com/questions/72396387/html-canvas-draw-smooth-lines-with-bezier-curve


But not to write too much.  I'll test it and come back.

Josip

**Comment by MarvinKlein1508** [comment]
@JosipNizetic this doesn't work becuase you wouldn't be able to change signatures anymore which are being inititalized by `OnParametersSetAsync`

We cannot clear the canvas for this reason. This would only be possible if the previous set value would be applied after the clearing

**Comment by MarvinKlein1508** [comment]
@CodeGod911 we should take a look on this for the other `lineCap`  & `lineJoin` options.

Drawing on `"lineCap": "round"` and `"lineJoin": "round"` results in much worse anti-aliasling

**Comment by CodeGod911** [comment]
> @JosipNizetic this doesn't work becuase you wouldn't be able to change signatures anymore which are being inititalized by `OnParametersSetAsync`
> 
> We cannot clear the canvas for this reason. **This would only be possible if the previous set value would be applied after the clearing**

Do you think it would be feasible to do?

**Comment by CodeGod911** [comment]
@JosipNizetic @MarvinKlein1508  here is an unfinshed approach what do you think https://github.com/CodeGod911/SignaturePad/tree/bugfix/Antialiasing

**Comment by CodeGod911** [comment]
> @JosipNizetic @MarvinKlein1508 here is an unfinshed approach what do you think https://github.com/CodeGod911/SignaturePad/tree/bugfix/Antialiasing

I havent looked at the line options yet

**Comment by JosipNizetic** [comment]
I am sorry , went on trip. I will look at it tonight. 

Joisip

**Comment by MarvinKlein1508** [comment]
@CodeGod911 anti-aliasing is much better now. The BGImage demo seem to be broken though.

**Comment by CodeGod911** [comment]
here is a second approach redrawing backgrounds https://github.com/CodeGod911/SignaturePad/tree/bugfix/Antialiasing2
the only issue i found with this is a minimal disapearance of the last line as long as blazor transfers the images back and forth. For me this is about 116ms. It only happens though if the drawing starts again within those 116ms which in general shouldn be a problem but theres no knowing with ping and datarates what could happen...

**Comment by MarvinKlein1508** [comment]
I try to test it tomorrow. Have much todo right now.

**Comment by MarvinKlein1508** [comment]
@CodeGod911 I've taken a look on version 2. The abckground image loads fine now. As you've pointed out, there is still an issue when the initial loading takes a while. This can easily be re-created by this steps:

1. Start project in Firefox
2. Press CTRL+SHIFT+M and slow down the internet connection to GPRS
3. Go to the background demo
4. draw while image is loading
5. See signature without background image


We should prevent drawing while the signature is still initializing the background image.

**Comment by MarvinKlein1508** [comment]
Fixed in v8.1.4

**Comment by JosipNizetic** [comment]
Everything is working flawlessly. Congratulations on a good job!
